### PR TITLE
Improve debugger function

### PR DESCRIPTION
**Pull request by ksh8281**
* Implement port option
* Set 'this' value correctly for eval in debugger
* Implement eval without stopping state
* Implement pumpDebuggerEvents function
* Remove useless variable m_thisExpressionIndex in class context information

Signed-off-by: Seonghyun Kim <sh8281.kim@samsung.com>

### CODE REVIEW

**Review by zherczeg — COMMENTED** [review]
The aim of the patch is evaluating scripts without stopping the execution, am I right? Are these scripts evaluated in the global context?

File: src/debugger/DebuggerTcp.cpp [hunk 1]
```diff
@@ -327,7 +354,11 @@ void DebuggerTcp::init(const char*, Context* context)
         return;
     }
 
-    printf("Waiting for client connection\n");
+    if (port == 6501) {
+        ESCARGOT_LOG_INFO("Waiting for client connection\n");
+    } else {
+        ESCARGOT_LOG_INFO("Waiting for client connection 0.0.0.0:%hd\n", port);
```
[the remark this answers is not in the document]
Comment (ksh8281) [reply src/debugger/DebuggerTcp.cpp:360]: Okay. :)

File: src/debugger/Debugger.h [hunk 2]
```diff
@@ -248,6 +255,11 @@ class DebuggerRemote : public Debugger {
         ESCARGOT_DEBUGGER_THERE_WAS_NO_SOURCE = 18,
         ESCARGOT_DEBUGGER_PENDING_CONFIG = 19,
         ESCARGOT_DEBUGGER_PENDING_RESUME = 20,
+        // These four must be in the same order.
+        ESCARGOT_MESSAGE_EVAL_WITHOUT_STOP_8BIT_START = 21,
+        ESCARGOT_MESSAGE_EVAL_WITHOUT_STOP_8BIT = 22,
+        ESCARGOT_MESSAGE_EVAL_WITHOUT_STOP_16BIT_START = 23,
+        ESCARGOT_MESSAGE_EVAL_WITHOUT_STOP_16BIT = 24,
```
[the remark this answers is not in the document]
Comment (ksh8281) [reply src/debugger/Debugger.h:262]: 1. Our python client could not support this function. but vs-code-extension can support this. I will add the support on vs-code-extension
2. If you think we are not maintaining a strict versioning, I can move these variables after `ESCARGOT_MESSAGE_EVAL_8BIT_START`

**Comment by ksh8281** [comment]
> The aim of the patch is evaluating scripts without stopping the execution, am I right? Are these scripts evaluated in the global context?

Both are Yes. It should be executed on global context, like chrome inspector

File: src/runtime/Context.cpp [hunk 3]
```diff
@@ -165,6 +165,30 @@ String* Context::getClientSource(String** sourceName)
...
+    if (m_debugger != nullptr) {
+        return;
+    }
+
+    m_debugger->setAsAlwaysStopState();
+}
+
+void Context::pumpDebuggerEvents()
+{
+    if (m_debugger != nullptr) {
```
Comment (zherczeg) [inline src/runtime/Context.cpp:179]: Isn't this a `m_debugger == nullptr`?

File: src/debugger/DebuggerTcp.cpp [hunk 4]
```diff
@@ -301,10 +302,36 @@ static bool webSocketHandshake(EscargotSocket socket)
...
+}
+
+void DebuggerTcp::init(const char* options, Context* context)
 {
     uint16_t port = 6501;
 
+    if (options) {
+        auto v = split(options, ';');
+        for (size_t i = 0; i < v.size(); i++) {
+            const std::string& s = v[i];
```
[the remark this answers is not in the document]
Comment (ksh8281) [reply src/debugger/DebuggerTcp.cpp:325]: I just want to use utility function `split` in String.h
I am going to fix my PR :(

File: src/runtime/Context.cpp (hunk 3 above)
Comment (ksh8281) [reply src/runtime/Context.cpp:179]: You are right. It is mistake!

**Review by zherczeg — APPROVED** [review]
(informal) LGTM

**Review by clover2123 — APPROVED** [review]
LGTM 👍

**Pull request merged**